[STEP-8] Event 활용해서 부가 로직 분리

n3rno opened the pull request:
## PR 설명
- Event 활용하여 주문완료 정보 외부 전송 기능을 분리
- `EcommerceException` 클래스 생성하여 예외처리 개선


## Definition of Done (DoD)
- [x] 실시간 주문정보를 데이터 플랫폼에 전송(mock API 호출)하는 요구사항을 이벤트를 활용하여 트랜잭션과 관심사를 분리한다.
- [ ] MSA 구조 도메인 설계 및 보고서 작성

n3rno committed:
Exception 클래스 생성 및 에러코드 변경

n3rno committed:
이벤트로 주문 데이터 외부 전송 기능 분리

Isaac482763 commented on src/main/java/kr/hhplus/be/server/order/application/service/OrderService.java:
```diff
@@ -66,19 +72,44 @@ public void order(OrderRequestDto orderRequestDto) {
+        }
+
+            // 주문 이력 생성
+            orderRepository.insertOrder(order);
+            orderRepository.insertOrderGoods(OrderGoods.from(orderId, orderRequestDto.getOrderGoodsList()));
```
주문 프로세스의 원자성을 보장하기 위해서는
주문 이력 생성 코드도 try문 안에 있어야 할 것 같습니다.
주문이력 생성 시점에 예기치 않은 오류가 발생하는 경우
이미 차감된 포인트와 재고는 롤백되지않아 데이터 불일치가 발생합니다.

Isaac482763 commented on src/main/java/kr/hhplus/be/server/order/application/service/OrderEventHandler.java:
```diff
@@ -0,0 +1,21 @@
+@Component
+public class OrderEventHandler {
+
+    @EventListener
+    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
```
```suggestion
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
```

트랜잭션 커밋 이후에 이벤트를 받아서 처리하도록 잘 하셨습니다!
다만 여기에서 @TransactionalEventListener 어노테이션이 @EventListener을 포함하므로 제거해도 될거같습니다.

Isaac482763 commented on src/main/java/kr/hhplus/be/server/order/application/service/OrderEventHandler.java:
```diff
@@ -0,0 +1,21 @@
+        System.out.printf("[MockSend] 주문 완료 정보 전송: orderId=%s, userNo=%d, amount=%d%n",
+                event.getOrderId(), event.getUserNo(), event.getTotalAmount());
+
+        // 실제 구현 시 WebClient, RestTemplate 등으로 API 호출
```
외부 API 호출을 추가하게되면
아래 예시처럼 API 호출 실패시에 재시도를 할수있게
retry 로직을 추가하면 좋을것같습니다.

```
    @TransactionalEventListener(phase = TransactionPhase.AFTER_COMMIT)
    @Retryable(
        value = RuntimeException.class,
        maxAttempts = 3,
        backoff = @Retryable.Backoff(delay = 2000) // 2초 간격 재시도
    )
    public void handleOrderCompleted(OrderCompletedEvent event) {
        // 외부 데이터 플랫폼으로 전송 (Mock API 호출)
        System.out.printf("[MockSend] 주문 완료 정보 전송: orderId=%s, userNo=%d, amount=%d%n",
                event.getOrderId(), event.getUserNo(), event.getTotalAmount());
        throw new RuntimeException("Mock API 전송 실패"); // 외부 API 호출시에 예외 던지도록 추가
        // 실제 구현 시 WebClient, RestTemplate 등으로 API 호출
    }
```

Isaac482763 commented on src/test/java/kr/hhplus/be/server/order/OrderTest.java:
```diff
@@ -40,6 +45,9 @@ public class OrderTest {
     @Mock
     private UserService userService;
+
+    @Mock
+    private ApplicationEventPublisher applicationEventPublisher;
```
```suggestion
    @MockBean
    private ApplicationEventPublisher applicationEventPublisher;
```
@Mock은 단순 Mockito Mock 객체를 생성하지만 Spring Context에 등록되지 않아서 @Autowired나 의존성 주입 대상에 적용되지 않습니다.
@MockBean은 Spring 테스트 컨텍스트에 해당 Mock을 주입해 실제 빈을 대체하므로 통합 테스트에서 의존성 주입이 제대로 작동하려면 @MockBean을 사용해야 합니다.

Isaac482763 commented on src/main/java/kr/hhplus/be/server/Exception/EcommerceException.java:
```diff
@@ -0,0 +1,21 @@
+
+@Getter
+@Setter
+public class EcommerceException extends RuntimeException {
```
커스텀 예외를 추가해서 훨씬 명확해진것 같습니다 ㅎㅎ

Isaac482763 commented on src/main/java/kr/hhplus/be/server/Exception/EcommerceException.java:
```diff
@@ -0,0 +1,21 @@
+    private String message;
+
+    public EcommerceException(ErrorCode errorCode) {
+        this.code = errorCode.getCode();
+        this.message = errorCode.getMessage();
```
부모 클래스에도 message값을 설정해주면 스택 트레이스 로깅에 유용합니다.
```suggestion
        super(errorCode.getMessage());
        this.code = errorCode.getCode();
        this.message = errorCode.getMessage();
```

Isaac482763 reviewed (approved):
안녕하세요 이번주도 고생하셨습니다!
이벤트 기반의 아키텍처를 잘 도입하셨습니다.

n3rno committed:
8주차 코멘트 수정사항 반영

1. 외부 api 호출 후 재시도 로직 추가
2. 주문이력 생성 try 내부로 이동
3. @Mock 어노테이션 @MockitoBean으로 변경
